[DCP] Add option to use PrefixStore to create checkpoint background process

kevinmtang opened the pull request:
Summary:
DCP checkpoint background process currently determines the port used for pg via get_free_port().

During checkpoint background process initialization, gloo pg init occasionally times out on the first call but succeeds in a subsequent call.

We hypothesized that the timeouts are related to the port being used, and the solution would be to create the pg with PrefixStore and reuse the master port.

This diff adds the option for checkpoint background process to use PrefixStore with MASTER_ADDR + MASTER_PORT.

The default behavior is unchanged. Enabling the new PrefixStore behavior requires setting "DCP_USE_PREFIX_STORE" env var to "1".

context:
 https://fb.workplace.com/groups/319878845696681/permalink/1516883985996155/

Differential Revision: D84928180




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

meetv18 reviewed (approved):
Thanks for adding support for this and improving reliability of process based async cp :)

kevinmtang committed:
[DCP] Add option to use PrefixStore to create checkpoint background process (#166560)

Summary:

DCP checkpoint background process currently determines the port used for pg via get_free_port().

During checkpoint background process initialization, gloo pg init occasionally times out on the first call but succeeds in a subsequent call.

We hypothesized that the timeouts are related to the port being used, and the solution would be to create the pg with PrefixStore and reuse the master port.

This diff adds the option for checkpoint background process to use PrefixStore with MASTER_ADDR + MASTER_PORT.

The default behavior is unchanged. Enabling the new PrefixStore behavior requires setting "DCP_USE_PREFIX_STORE" env var to "1".

context:
 https://fb.workplace.com/groups/319878845696681/permalink/1516883985996155/

Test Plan:
```
buck test fbcode//mode/opt fbcode//caffe2/test/distributed/checkpoint:test_async_process_executor -- --run-disabled
```
https://www.internalfb.com/intern/testinfra/testrun/1407375340702140

https://www.internalfb.com/mlhub/pipelines/runs/mast/aps-omnifm_0928_6kx_16-ETT-1348807470?job_attempt=1&version=0&tab=summary&env=PRODUCTION

https://www.internalfb.com/mlhub/pipelines/runs/mast/aps-omnifm_0928_6kx_256-ETT-28070b842e?version=0&env=PRODUCTION

Reviewed By: meetv18

Differential Revision: D84928180
